src/libs/Middleware/Pagination.ts

```diff
@@ -0,0 +1,127 @@
+// TODO: Is this a legit use case for exposing `OnyxCache`, or should we use `Onyx.connect`?
+import fastMerge from 'expensify-common/dist/fastMerge';
+import type {OnyxCollection} from 'react-native-onyx';
+import Onyx from 'react-native-onyx';
+import type {ApiCommand} from '@libs/API/types';
+import Log from '@libs/Log';
+import PaginationUtils from '@libs/PaginationUtils';
+import CONST from '@src/CONST';
+import type {OnyxCollectionKey, OnyxPagesKey, OnyxValues} from '@src/ONYXKEYS';
+import type {Request} from '@src/types/onyx';
+import type {PaginatedRequest} from '@src/types/onyx/Request';
+import type Middleware from './types';
+
+type PaginationConfig<TResourceKey extends OnyxCollectionKey, TPageKey extends OnyxPagesKey> = {
+    initialCommand: ApiCommand;
+    previousCommand: ApiCommand;
+    nextCommand: ApiCommand;
+    resourceCollectionKey: TResourceKey;
+    pageCollectionKey: TPageKey;
+    sortItems: (items: OnyxValues[TResourceKey]) => Array<OnyxValues[TResourceKey] extends Record<string, infer TResource> ? TResource : never>;
+    getItemID: (item: OnyxValues[TResourceKey] extends Record<string, infer TResource> ? TResource : never) => string;
```

Comment:
I don't understand the logic with these two types here, why not simplify it a bit?
```ts
    sortItems: (items: OnyxValues[TResourceKey]) => Array<OnyxValues[TResourceKey]>;
    getItemID: (item: OnyxValues[TResourceKey]) => string;
```

Reply:
Also shouldn't it be `(items: OnyxValues[TResourceKey])` in `sortItems`?

Reply:
The tricky thing with this type is that what we're paginating isn't actually `OnyxValues[ONYXKEYS.COLLECTION.REPORT_ACTIONS]`. because the ReportActions collection is different from others.

For example, with the Report collection, you have something like this:

```
report_1234 => Report,
report_4567 => Report,
report_${reportID} => Report,
```

and you'd want to paginate over the `report_` keys. Whereas with `ReportActions`, it's like this:

```
reportActions_${reportID} => {
    ${reportActionID} => ReportAction;
};
```

and for ReportActions you want to paginate over the collection stored within a single key

Reply:
Figured out some utility types that might be helpful:

```ts

type ExtractPrefix<TStr extends string, TSuffix extends string> = TStr extends `${infer Prefix}${TSuffix}` ? Prefix : never;
type ExtractOnyxCollectionKey<TKey extends OnyxCollectionKey | `${OnyxCollectionKey}${string}`> = TKey extends OnyxCollectionKey ? TKey : `${ExtractPrefix<TKey, `_${string}`>}_`;
```

Reply:
I think we may need to do cleanup here in the future if/when we want to reuse this for report pagination. But in the meantime, this works ok. Let's leave it as-is.